sympy/matrices/expressions/tests/test_derivatives.py

```diff
@@ -86,3 +86,106 @@ def test_matrix_derivative_with_inverse():
     # Cookbook example 64:
     expr = Trace(Inverse(X + A))
     assert expr.diff(X) == -(Inverse(X + A)*Inverse(X + A)).T
+
+
+def test_matrix_derivatives_of_traces():
+
+    ## First order:
+
+    # Cookbook example 99:
+    expr = Trace(X)
+    assert expr.diff(X) == Identity(k)
+
+    # Cookbook example 100:
+    expr = Trace(X*A)
+    assert expr.diff(X) == A.T
+
+    # Cookbook example 101:
+    expr = Trace(A*X*B)
+    assert expr.diff(X) == A.T*B.T
+
+    # Cookbook example 102:
+    expr = Trace(A*X.T*B)
+    assert expr.diff(X) == B*A
+
+    # Cookbook example 103:
+    expr = Trace(X.T*A)
+    assert expr.diff(X) == A
+
+    # Cookbook example 104:
+    expr = Trace(A*X.T)
+    assert expr.diff(X) == A
+
+    # Cookbook example 105:
+    # TODO: TensorProduct is not supported
+    #expr = Trace(TensorProduct(A, X))
+    #assert expr.diff(X) == Trace(A)*Identity(k)
+
+    ## Second order:
+
+    # Cookbook example 106:
+    expr = Trace(X**2)
+    assert expr.diff(X) == 2*X.T
+
+    # Cookbook example 107:
+    expr = Trace(X**2*B)
+    # TODO: wrong result
+    #assert expr.diff(X) == (X*B + B*X).T
```

Comment:
@asmeurer a new bug has emerged. I guess we need more extensive testing (i.e. test all rules by the cookbook).

Reply:
OK. Do you plan to fix this here, or should this be merged as is?

Reply:
For reference, in master, this raises `TypeError: Mix of Matrix and Scalar symbols`.

Reply:
https://github.com/sympy/sympy/issues/15136